app/models/user.rb

```diff
@@ -77,7 +77,7 @@ class User < ActiveRecord::Base
                    :length => { :maximum => 254 },
                    :allow_blank => true
   validates :mail, :presence => true, :on => :update,
-                   :if => Proc.new { |u| !AuthSourceHidden.where(:id => u.auth_source_id).any? && u.mail_was.present? }
+                   :if => Proc.new { |u| !AuthSourceHidden.where(:id => u.auth_source_id).any? && (u.mail_was.present? || User.current.id == u.id && !User.current.hidden?) }
```

Comment:
Couple of small nitpicks:

1. Use brackets around the second `(User.current.id == u.id && !User.current.hidden?)` so we're not relying entirely on precedence.
2. `User.current == u` should work, and is shorter.